d2/src/main/java/com/linkedin/d2/balancer/servers/ZooKeeperAnnouncer.java

```diff
@@ -195,26 +233,18 @@ public ZooKeeperAnnouncer(LoadBalancerServer server, boolean initialIsUp,
   public ZooKeeperAnnouncer(ZooKeeperServer server, boolean initialIsUp,
                             boolean isDarkWarmupEnabled, String warmupClusterName, int warmupDuration, ScheduledExecutorService executorService, ServiceDiscoveryEventEmitter eventEmitter)
   {
-    _server = server;
-    // initialIsUp is used for delay mark up. If it's false, there won't be markup when the announcer is started.
-    _isUp = initialIsUp;
-    _isWarmingUp = false;
-    _isRetryWarmup = false;
-    _pendingMarkDown = new ArrayDeque<>();
-    _pendingMarkUp = new ArrayDeque<>();
-    _pendingWarmupMarkDown = new ArrayDeque<>();
-
-    _isDarkWarmupEnabled = isDarkWarmupEnabled;
-    _warmupClusterName = warmupClusterName;
-    _warmupDuration = warmupDuration;
-    _executorService = executorService;
-    _eventEmitter = eventEmitter;
-
-    server.setServiceDiscoveryEventHelper(this);
+    this(server, initialIsUp, isDarkWarmupEnabled, warmupClusterName, warmupDuration, executorService, eventEmitter, null, ActOnWeightBreach.IGNORE);
   }
 
   public ZooKeeperAnnouncer(LoadBalancerServer server, boolean initialIsUp,
       boolean isDarkWarmupEnabled, String warmupClusterName, int warmupDuration, ScheduledExecutorService executorService, ServiceDiscoveryEventEmitter eventEmitter)
+  {
+    this(server, initialIsUp, isDarkWarmupEnabled, warmupClusterName, warmupDuration, executorService, eventEmitter, null, ActOnWeightBreach.IGNORE);
+  }
+
+  public ZooKeeperAnnouncer(LoadBalancerServer server, boolean initialIsUp,
+      boolean isDarkWarmupEnabled, String warmupClusterName, int warmupDuration, ScheduledExecutorService executorService,
+      ServiceDiscoveryEventEmitter eventEmitter, String maxWeight, ActOnWeightBreach actOnWeightBreach)
```

Comment:
Why not take the `maxWeight` as a `BigDecimal` directly and let callers figure out what they want?

Reply:
synced offline. I was thinking of using string to help callers less likely to mis-configure the decimal places on max weight (if people write new BigDecimal(100), it will have 1 decimal places, while new BigDecimal("100.00") will precisely define 2 decimal places.). But actually string won't help much either if people don't pay close attention to how to define it. Updated to BigDecimal.